text/0000-fabric-gateway.md

```diff
@@ -0,0 +1,213 @@
+[guide-level-explanation]: #guide-level-explanation
+
+The Fabric Gateway is an embodiment of the high-level 'gateway' Fabric programming model in a server component that will form part of a Fabric network alongside Peers, Orderers and CAs.  It will either be stood up inside its own process in a docker container, or it will be hosted inside a peer.  Either way, it exposes its functionality to clients via a gRPC interface.
+
+Lightweight client SDKs are used to connect to the Gateway for the purpose of invoking transactions.  The gateway will intereact with the rest of the network on behalf of the client eliminating the need for the client application to connect directly to peers and orderers.
+
+The concepts and APIs will be familiar to Fabric programmers who are already using the new programming model.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The Gateway runs as a client process associated with an organisation and requires an identity (issued by the org's CA) in order to interact with the discovery service and to register for events (Deliver Client).  
+
+The Gateway also runs as a gRPC server and exposes the following services to client applications (SDKs):
+
+```
+service Gateway {
+    rpc Endorse(ProposedTransaction) returns (PreparedTransaction) {}
+    rpc Submit(PreparedTransaction) returns (stream Event) {}
+    rpc Evaluate(ProposedTransaction) returns (Result) {}
+}
+```
+
+The client credentials (private key) are never passed to the Gateway.  Client applications are responsible for managing their user keys (optionally using SDK Wallets) and signing the protobuf payloads.
+
+Submitting a transaction is a two step process (performed by the client SDK):
+- __Endorse__
+  - The client creates a ProposedTransaction message, which contains a SignedProposal message as defined in `fabric-protos/peer/proposal.proto` and signed with the user's identity.  The ProposedTreansaction message also contains other optional properties, such as the endorsing orgs if the client wants to specify that.
+  - The `Endorse` service is invoked on the Gateway, passing the ProposedTransaction message
+    - The Gateway will determine the endorsement plan for the requested chaincode and forward to the appropriate peers for endorsement. It will return to the client a `PreparedTransaction` message which contains a `Envelope` message as defined in `fabric-protos/common/common.proto`.  It will also contain other information, such as the return value of the chaincode function and the transaction ID so that the client doesn't necessarily need to unpack the `Envelope` payload.
+- __Submit__
+  - The client signs the hash of the Envelope payload using the user's private key and sets the signature field in the `Envelope` in the `PreparedTransaction`.
+  - The `Submit` service is invoked on the Gateway, passing the signed `PreparedTransaction` message.  A stream is opened to return multiple return values.
```

Comment:
The stream interaction is proposed to allow the client to:
1. Send the transaction to Fabric Gateway.
1. Get a response and do follow-on work when the transaction is successfully delivered to the orderer.
1. Subsequently wait for Fabric Gateway to report successful commit of the transaction.

How would this flow work without streaming two separate response messages back from Fabric Gateway to the client?

Reply:
I think I agree with Bret. Why do we need to separate 2 and 3? I'd expect a client to want a single response that indicates the transaction disposition. Whether or not that operation is handled asynchronously should be a detail of the caller and not the API.

Reply:
@sykesm The existing fabric programming model in the client SDKs has a single return point from a transaction submit which, by default, returns the transaction result (or an error) to the client application once the transaction has been committed (or the commit was unsuccessful). So the existing SDKs behaves as you describe:
```
result, err := contract.SubmitTransaction(txName, args...)
```

We have had numerous instances of client application developers wanting to:
1. Receive an acknowledgement when the transaction is successfully delivered to the orderer; then subsequently
1. Receive the transaction commit disposition.

Reasons cited for this requirement are that they want to:
- Provide a more responsive UI, where it is updated immediately after delivery to the orderer but can still coordinate follow-on work and/or UI updates after transaction commit.
- Where the Fabric client application is itself a REST server, to provide a more rapid RESTful response to their downstream clients immediately on delivery to the orderer as they might otherwise result in timed out connections, but still handle commit disposition.

Today the only way for client applications to achieve this behaviour is to do the transaction submit in a mode that does not wait for the transaction commit and returns the transaction result immediately after the transaction is delivered to the orderer. They then need to implement their own commit notification using eventing. This is significantly more work for the client application and requires the client to deal with the mechanics of transaction commits.

The intention of this streamed two-stage response from a transaction submit is to:
1. Provide a significantly simpler client application development experience where this decoupled behaviour is desired.
2. Allow commit handling to be implemented entirely on the server side of the connection, where it can be carried out much closer to the Fabric nodes (for improved performance) and in a single language (Go) rather than in multiple client-side languages.

Bear in mind that this streamed two-stage response is implemented at the gRPC level. Client applications are typically developed to the client-side SDK. In the client-side SDK, we can still expose the transaction invocation as a single method call that returns only on commit, in addition to a form that allows a two-stage response. So for the client application there is zero additional complexity associated with a streaming gRPC response. For example, something along these lines:
```
// Returns result only after commit
result, err := contract.SubmitTransaction(txName, args...)

// Returns result on delivery to orderer and signals subsequent commit on the returned channel
result, commitChannel, err := contract.SubmitAsync(txName, args...)
```

If we can achieve the same client application development simplicity and experience without using a streaming gRPC response, and also keep the commit handling implementation server-side, that would be great. How can we do that?

Reply:
My perspective is that the suggested approach is on the cusp of abusing streamed responses. A gRPC stream response is generally intended to iterate over a result set or deliver a stream of events. In this case, it looks like the event stream is actually one event (completion).

Unless I'm misreading, the response from Submit can be delayed until the transaction has been submitted (meeting the first requirement) and the client can then use the response to implement a wait for completion.

If `Event` actually means _n_ events are going to come back from the gateway where _n_ is generally greater than 2, I may not have the same reaction.

Reply:
To conclude this discussion (including an offline discussion with @sykesm), we have decided to stay with the proposed Submit function that returns a stream.  In the spirit of agile development, if implementation experience requires us to revisit this, then we can do so before this component is formally released.